### ISSUE TITLE
Batch inference of YOLO11 seg model with NMS in NVIDIA Triton is wrong.

### ISSUE THREAD

**Issue by kino0924** [ultralytics/ultralytics#23425]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Integrations, Other

### Bug

I could not batch anything larger than 2 because output0 is pretty much empty.

this is output of my test script with batch size of 1 (total 8 inference)
First column is hash of output0 and second column is hash of output1
```
batch_nms ready: True
batch_nms output0 shape: (1, 300, 38)
batch_nms output1 shape: (1, 32, 104, 136)
batch_nms output 0 hash: 329ed58a8762e9ff3f13c89eea2e6b48 - f746bf7b0b68675dcafb2757ce3ba673
batch_nms output 1 hash: 107c72102582bc199ab7824517359528 - c2813dec058976e84d6fe4556dd20d0a
batch_nms output 2 hash: fbb264d9d4ff8b5b174be015444ee12b - 00729b8851986d9ff885fa3e1fe52ae7
batch_nms output 3 hash: be786972df5e5f80f88d4741af2156d1 - ccfbde56b76ce1426fd1f77f8527398d
batch_nms output 4 hash: 860ccc65ef44974df672ba92d0231195 - 29bf1d250acd7ad18de9e42a5ba4ed05
batch_nms output 5 hash: b59ffcb424d1567378d803beb5929093 - 0a43d8f2273f235b9516a398367c919c
batch_nms output 6 hash: eee796492933fbbe2e277c21a4500e0c - 25138ab51c679c3d2a321dd6b079841c
batch_nms output 7 hash: 4bb1e1022d51a34315318ad4c4b1cc39 - 30ff9fe111898b171e997aae97a2bddf
```

Now, using same data, I did batch inference with batch size of 8 (total 1 inference)
```
batch_nms ready: True
batch_nms output0 shape: (8, 300, 38)
batch_nms output1 shape: (8, 32, 104, 136)
batch_nms output 0 hash: 329ed58a8762e9ff3f13c89eea2e6b48 - f746bf7b0b68675dcafb2757ce3ba673
batch_nms output 1 hash: 2434ea5e1e58e47d12f4d17b80dbd73d - c2813dec058976e84d6fe4556dd20d0a
batch_nms output 2 hash: 2434ea5e1e58e47d12f4d17b80dbd73d - 00729b8851986d9ff885fa3e1fe52ae7
batch_nms output 3 hash: 2434ea5e1e58e47d12f4d17b80dbd73d - ccfbde56b76ce1426fd1f77f8527398d
batch_nms output 4 hash: 2434ea5e1e58e47d12f4d17b80dbd73d - 29bf1d250acd7ad18de9e42a5ba4ed05
batch_nms output 5 hash: 2434ea5e1e58e47d12f4d17b80dbd73d - 0a43d8f2273f235b9516a398367c919c
batch_nms output 6 hash: 2434ea5e1e58e47d12f4d17b80dbd73d - 25138ab51c679c3d2a321dd6b079841c
batch_nms output 7 hash: 2434ea5e1e58e47d12f4d17b80dbd73d - 30ff9fe111898b171e997aae97a2bddf
```

I tried different version of Nvidia Triton as well but the results are same (currently on latest 2.64.0 and corresponds to the 25.12 container)

### Environment

```
Ultralytics 8.3.252 🚀 Python-3.10.12 torch-2.9.1+cu128 CUDA:0 (NVIDIA GeForce RTX 3090, 24134MiB)
Setup complete ✅ (64 CPUs, 345.8 GB RAM, 103.4/196.3 GB disk)

OS                     Linux-5.15.0-164-generic-x86_64-with-glibc2.35
Environment            Linux
Python                 3.10.12
Install                pip
Path                   /home/lack/.local/lib/python3.10/site-packages/ultralytics
RAM                    345.76 GB
Disk                   103.4/196.3 GB
CPU                    Intel Xeon Silver 4216 CPU @ 2.10GHz
CPU count              64
GPU                    NVIDIA GeForce RTX 3090, 24134MiB
GPU count              4
CUDA                   12.8

numpy                  ✅ 2.2.6>=1.23.0
matplotlib             ✅ 3.10.8>=3.3.0
opencv-python          ✅ 4.12.0.88>=4.6.0
pillow                 ✅ 12.1.0>=7.1.2
pyyaml                 ✅ 5.4.1>=5.3.1
requests               ✅ 2.25.1>=2.23.0
scipy                  ✅ 1.15.3>=1.4.1
torch                  ✅ 2.9.1>=1.8.0
torch                  ✅ 2.9.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.24.1>=0.9.0
psutil                 ✅ 7.2.1>=5.8.0
polars                 ✅ 1.37.1>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18

```

### Minimal Reproducible Example

model export.

```
# Load the YOLO11 model
model = YOLO("model.pt")
model.export(format="onnx", opset=18, device=0, dynamic=True, half=True, nms=True)
```

TensorRT conversion
```
/usr/src/tensorrt/bin/trtexec --onnx=/tmp/model.onnx --minShapes=images:1x3x416x416 --optShapes=images:1x3x416x544 --maxShapes=images:8x3x768x768 --saveEngine=/tmp/model_fp16_nms.plan --fp16
```

config.pbtxt
```
name: "batch_nms"
platform: "tensorrt_plan"
max_batch_size: 8

input [
  {
    name: "images"
    data_type: TYPE_FP16
    dims: [3, -1, -1]
  }
]

output [
  {
    name: "output0"
    data_type: TYPE_FP16
    dims: [300, 38]
  },
  {
    name: "output1"
    data_type: TYPE_FP16
    dims: [32, 192, 192]
  }
]
```

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @kino0924, thank you for your detailed report and for your interest in Ultralytics 🚀! This is an **automated response** to help us triage quickly—an Ultralytics engineer will also take a look and assist soon 🛠️

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this is a 🐛 Bug Report, please share a **minimum reproducible example (MRE)** (if possible, even smaller than the export snippet) using our guide here: [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/). In particular, an MRE that includes the exact Triton request code (inputs, batching, dtypes, and shapes) and how you’re hashing/validating `output0`/`output1` would help us reproduce the “batch > 2 becomes empty” behavior quickly 🔍

If this is related to custom training ❓ (or model-specific behavior), please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
This looks like an export-time batching issue: when `nms=True`, the ONNX NMS wrapper is traced with the export `batch` and the per-image NMS loop gets effectively “unrolled” to that size, so exporting with the default `batch=1` commonly results in only batch element 0 having valid NMS output when you run larger batches in TensorRT/Triton; please re-export with your max batch (e.g. 8), rebuild the `.plan`, and re-test (see the batching/padding logic in the [NMS exporter implementation](https://docs.ultralytics.com/reference/engine/exporter/)).

```python
from ultralytics import YOLO

model = YOLO("model.pt")
model.export(format="onnx", opset=18, device=0, dynamic=True, half=True, nms=True, batch=8)
```

Also double-check your Triton config matches the engine’s batching mode (ONNX→TensorRT is typically explicit-batch), e.g.:

```protobuf
name: "batch_nms"
platform: "tensorrt_plan"
max_batch_size: 0

input [{ name: "images" data_type: TYPE_FP16 dims: [-1, 3, -1, -1] }]
output [
  { name: "output0" data_type: TYPE_FP16 dims: [-1, 300, 38] }
  { name: "output1" data_type: TYPE_FP16 dims: [-1, 32, -1, -1] }
]
```

If it still reproduces after exporting with `batch=8`, please paste the `model.export(...)` console output (it prints the traced input/output shapes) and the exact Triton client request shapes you’re sending.

**Comment by Y-T-G** [comment]
There would have been a warning in the export logs too about this:

```
WARNING ⚠️ 'dynamic=True' model with 'nms=True' requires max batch size, i.e. 'batch=16'
```

**Comment by kino0924** [comment]
@glenn-jocher Thank you so much. Setting max batch number while export fixed the issue.
I thought giving bacth=x would strictly limit for exact batch size only but it also works on  batch <=8
 
```
from ultralytics import YOLO

# Load the YOLO11 model
model = YOLO("model.pt")
model.export(format="onnx", opset=18, device=0, dynamic=True, half=True, nms=True, batch=8)
```

```
Ultralytics 8.3.252 🚀 Python-3.12.12 torch-2.10.0+cu128 CUDA:0 (NVIDIA GeForce RTX 3090, 24124MiB)
YOLO11s-seg summary (fused): 113 layers, 10,067,203 parameters, 0 gradients, 32.8 GFLOPs

PyTorch: starting from 'model.pt' with input shape (8, 3, 768, 768) BCHW and output shape(s) ((8, 300, 38), (8, 32, 192, 192)) (38.9 MB)

ONNX: starting export with onnx 1.20.1 opset 18...
/home/user/miniforge3/lib/python3.12/site-packages/ultralytics/engine/exporter.py:1524: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.detach().clone() or sourceTensor.detach().clone().requires_grad_(True), rather than torch.tensor(sourceTensor).
  pad = torch.zeros(torch.max(torch.tensor(self.args.batch - bs), torch.tensor(0)), *pred.shape[1:], **kwargs)
ONNX: slimming with onnxslim 0.1.82...
ONNX: export success ✅ 3.4s, saved as 'model.onnx' (19.7 MB)

Export complete (3.9s)
Results saved to /home/user
Predict:         yolo predict task=segment model=model.onnx imgsz=768 half
Validate:        yolo val task=segment model=model.onnx imgsz=768 data=/workspace/object_ocr/yolo/dataset/data.yaml half
Visualize:       https://netron.app
```

@Y-T-G That Warning message did not pop up on me.

**Comment by Y-T-G** [comment]
> @Y-T-G That Warning message did not pop up on me.

It should've, because it's present in the version you're using. Using your command from the original post, it appears immediately after the Ultralytics version.

```
model.export(format="onnx", opset=18, device=0, dynamic=True, half=True, nms=True)

Ultralytics 8.3.252 🚀 Python-3.11.13 torch-2.10.0+cu128 CUDA:0 (NVIDIA A100-SXM4-80GB, 81154MiB)
WARNING ⚠️ 'dynamic=True' model with 'nms=True' requires max batch size, i.e. 'batch=16'
```
